scripting/src/main/java/org/infinispan/scripting/impl/DistributedRunner.java

```diff
@@ -0,0 +1,45 @@
+package org.infinispan.scripting.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Future;
+
+import javax.script.ScriptException;
+
+import org.infinispan.Cache;
+import org.infinispan.commons.util.concurrent.Futures;
+import org.infinispan.commons.util.concurrent.NotifyingFuture;
+import org.infinispan.commons.util.concurrent.NotifyingFutureImpl;
+import org.infinispan.distexec.DefaultExecutorService;
+import org.infinispan.scripting.SystemBindings;
+import org.infinispan.scripting.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+public class DistributedRunner implements ScriptRunner {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass(), Log.class);
+   public static final DistributedRunner INSTANCE = new DistributedRunner();
+
+   private DistributedRunner() {
+   }
+
+   @Override
+   public <T> Future<T> runScript(ScriptingManagerImpl scriptManager, String scriptName, ScriptMetadata metadata, CacheScriptBindings binding) throws ScriptException {
+      Cache<?, ?> masterCacheNode = (Cache<?, ?>) binding.get(SystemBindings.CACHE);
+      if (masterCacheNode == null) {
+         throw log.distributedTaskNeedCacheInBinding(scriptName);
+      }
+      DefaultExecutorService des = new DefaultExecutorService(masterCacheNode);
+      List<Future<T>> tasks = des.submitEverywhere(new DistributedScript<T>(scriptName));
+      des.shutdown();
+      List<NotifyingFuture<T>> notifyingTasks = new ArrayList<>(tasks.size());
+      for (Future<T> task : tasks) {
+         NotifyingFutureImpl<T> nf = new NotifyingFutureImpl<>();
+         nf.setFuture(task);
+         notifyingTasks.add(nf);
+      }
+      NotifyingFuture<Void> combinedFutures = Futures.combine(notifyingTasks);
+
+      return null; // FIXME
```

Comment:
Looks like we need a new Future composition method for this case:

``` public NotifyingFuture<List<NotifyingFuture<T>>> sequence(Iterable<NotifyingFuture<T>> futures) ``` 
instead of  ```Futures.combine(...)```

This composition is commonly found in Scala (```all()```), Java8(```sequence()```) and Guava(```sucessfullAsList()```), would that solve your problem?